pkg/oauth/apiserver/apiserver.go

```diff
@@ -95,49 +144,103 @@ func (c *completedConfig) newV1RESTStorage() (map[string]rest.Storage, error) {
 
-	oauthClient, err := oauthclient.NewForConfig(c.GenericConfig.LoopbackClientConfig)
-	if err != nil {
-		return nil, err
-	}
 	routeClient, err := routeclient.NewForConfig(c.kubeAPIServerClientConfig)
 	if err != nil {
-		return nil, err
-	}
-	coreV1Client, err := corev1.NewForConfig(c.kubeAPIServerClientConfig)
-	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 
 	combinedOAuthClientGetter := oauthserviceaccountclient.NewServiceAccountOAuthClientGetter(
-		coreV1Client,
-		coreV1Client,
-		coreV1Client.Events(""),
+		corev1Client,
+		corev1Client,
+		corev1Client.Events(""),
 		routeClient,
-		oauthClient.OAuthClients(),
+		oauthClient.OauthV1().OAuthClients(),
 		saAccountGrantMethod,
 	)
 	authorizeTokenStorage, err := authorizetokenetcd.NewREST(c.GenericConfig.RESTOptionsGetter, combinedOAuthClientGetter)
 	if err != nil {
-		return nil, fmt.Errorf("error building REST storage: %v", err)
+		return nil, nil, fmt.Errorf("error building REST storage: %v", err)
 	}
 	accessTokenStorage, err := accesstokenetcd.NewREST(c.GenericConfig.RESTOptionsGetter, combinedOAuthClientGetter)
 	if err != nil {
-		return nil, fmt.Errorf("error building REST storage: %v", err)
+		return nil, nil, fmt.Errorf("error building REST storage: %v", err)
 	}
 	clientAuthorizationStorage, err := clientauthetcd.NewREST(c.GenericConfig.RESTOptionsGetter, combinedOAuthClientGetter)
 	if err != nil {
-		return nil, fmt.Errorf("error building REST storage: %v", err)
+		return nil, nil, fmt.Errorf("error building REST storage: %v", err)
 	}
 	userOAuthAccessTokensDelegate, err := useroauthaccesstokensdelegate.NewREST(accessTokenStorage)
 	if err != nil {
-		return nil, fmt.Errorf("error building REST storage: %v", err)
+		return nil, nil, fmt.Errorf("error building REST storage: %v", err)
 	}
+	tokenReviewStorage, tokenReviewPostStartHooks, err := c.tokenReviewStorage(corev1Client, oauthClient, userClient)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error building REST storage: %v", err)
+	}
+
+	v1Storage := map[string]rest.Storage{
+		"oAuthAuthorizeTokens":      authorizeTokenStorage,
+		"oAuthAccessTokens":         accessTokenStorage,
+		"oAuthClients":              clientStorage,
+		"oAuthClientAuthorizations": clientAuthorizationStorage,
+		"userOAuthAccessTokens":     userOAuthAccessTokensDelegate,
+		"tokenReviews":              tokenReviewStorage,
+	}
+	return v1Storage, tokenReviewPostStartHooks, nil
+}
+
+func (c *completedConfig) tokenReviewStorage(
+	corev1Client corev1.CoreV1Interface,
+	oauthClient *oauthclients.Clientset,
+	userClient *userclient.Clientset,
+) (rest.Storage, map[string]genericapiserver.PostStartHookFunc, error) {
+	bootstrapUserDataGetter := bootstrap.NewBootstrapUserDataGetter(corev1Client, corev1Client)
+
+	// add indexes to the userinformer for the users to be used in user <-> groups mapping
+	userInformer := c.ExtraConfig.UserInformers
+	if err := userInformer.User().V1().Groups().Informer().AddIndexers(cache.Indexers{
```

Comment:
move this up where you initialize the informers.  this extra index is pretty integral.